src/coreclr/jit/importer.cpp

```diff
@@ -9839,11 +9841,56 @@ void Compiler::impImportBlockCode(BasicBlock* block)
 #endif
                 {
                     /* Create a call to 'new' */
+                    helper = info.compCompHnd->getNewArrHelper(resolvedToken.hClass);
 
                     // Note that this only works for shared generic code because the same helper is used for all
                     // reference array types
-                    op1 =
-                        gtNewHelperCallNode(info.compCompHnd->getNewArrHelper(resolvedToken.hClass), TYP_REF, op1, op2);
+                    op1 = gtNewHelperCallNode(helper, TYP_REF, op1, op2);
+                }
+
+                // If we're jitting a static constructor and detect the following code pattern:
+                //
+                //  newarr
+                //  stsfld
+                //  ret
+                //
+                // We replace default heap allocator for newarr with the one that prefers frozen segments.
+                // This is a very simple and conservative implementation targeting Array.Empty<T>(), ideally
+                // we want to be able to use frozen allocators more broadly, but that analysis is not trivial.
+                //
+                if (((info.compFlags & FLG_CCTOR) == FLG_CCTOR) &&
+                    // NativeAOT is able to preinitialize objects on frozen segments without our help
```

Comment:
It would be better to clear `JIT_FLAG_FROZEN_ALLOC_ALLOWED` flag on the EE side.